microservices/gatewayApi/v2/routes/consumers.py

```diff
@@ -0,0 +1,472 @@
+    selectTag = "ns.%s" % namespace
+    ns_qualifier = None
+
+    if plugin_data == None:
+        log.error("Missing input")
+        record_custom_event(event_id, 'GatewayConsumerPlugin', 'add', 'failed', namespace, "Missing input.")
+        abort(make_response(jsonify(error="Missing input."), 400))
+
+    transform_tags(plugin_data, namespace, "ns.%s" % namespace)
+
+    if plugin_data['name'] == 'rate-limiting':
+        rate_limiting(plugin_data)
+
+    validate_tags(plugin_data, selectTag)
+
+    nsq = traverse_get_ns_qualifier(plugin_data, selectTag)
+    if nsq is not None:
+        if ns_qualifier is not None and nsq != ns_qualifier:
+            abort_early(event_id, 'GatewayConsumerPlugin', namespace, jsonify(error="Validation Errors:\n%s" %
+                        ("Conflicting ns qualifiers (%s != %s)" % (ns_qualifier, nsq))))
+        ns_qualifier = nsq
+        log.info("[%s] CHANGING ns_qualifier %s" % (namespace, ns_qualifier))
+
+    if ns_qualifier is not None:
+        selectTag = ns_qualifier
+
+    try:
+        response = cnsr.add_consumer_plugin(consumer_id, plugin_data)
+    except Exception as ex:
+        handle_exceptions(exception=ex, event_id=event_id, type='GatewayConsumerPlugin',
+                          action='add', namespace=namespace, message='failed to create plugin.')
+
+    record_custom_event(event_id, 'GatewayConsumerPlugin', 'add', 'completed', namespace)
+    return response
+
+
+@admin_jwt(None)
+@uma_enforce('namespace', 'GatewayConfig.Publish')
+@consumers.route('/<string:consumer_id>/plugins/<string:plugin_id>',
+                 methods=['PUT'], strict_slashes=False)
+def update_consumer_plugin(namespace: str, consumer_id: str, plugin_id: str) -> object:
+    event_id = str(uuid.uuid4())
+    record_custom_event(event_id, 'GatewayConsumerPlugin', 'update', 'received', namespace)
+    log = app.logger
+    plugin_data = request.get_json()
+    cnsr = GatewayConsumerService()
+    selectTag = "ns.%s" % namespace
+    ns_qualifier = None
+
+    if plugin_data == None:
+        log.error("Missing input")
+        record_custom_event(event_id, 'GatewayConsumerPlugin', 'update', 'failed', namespace, "Missing input.")
+        abort(make_response(jsonify(error="Missing input."), 400))
+
+    transform_tags(plugin_data, namespace, "ns.%s" % namespace)
+
+    if plugin_data['name'] == 'rate-limiting':
+        rate_limiting(plugin_data)
+
+    validate_tags(plugin_data, selectTag)
+
+    nsq = traverse_get_ns_qualifier(plugin_data, selectTag)
+    if nsq is not None:
+        if ns_qualifier is not None and nsq != ns_qualifier:
+            abort_early(event_id, 'GatewayConsumerPlugin', namespace, jsonify(error="Validation Errors:\n%s" %
+                        ("Conflicting ns qualifiers (%s != %s)" % (ns_qualifier, nsq))))
+        ns_qualifier = nsq
+        log.info("[%s] CHANGING ns_qualifier %s" % (namespace, ns_qualifier))
+
+    if ns_qualifier is not None:
+        selectTag = ns_qualifier
+
```

Comment:
I think we need to retrieve the plugin first to make sure that it has the proper `tag` related to the `ns_qualifier`, otherwise it is possible to overwrite a plugin that you don't technically have permission to change.  Same goes for Delete.

Reply:
@ikethecoder added a new function to verify the tags during update and delete.